confgenerator/confgenerator_test.go

```diff
@@ -48,94 +46,119 @@ var (
+		defaultConfig:   "windows-default-config.yaml",
+		InfoStat: &host.InfoStat{
+			OS:              "windows",
+			Platform:        "win_platform",
+			PlatformVersion: "win_platform_version",
+		},
+	},
+}
 
-			// Retrieve the expected golden conf files.
-			expectedMainConfig := readFileContent(testName, goldenMainPath, t, true)
-			expectedParserConfig := readFileContent(testName, goldenParserPath, t, true)
-			// Generate the actual conf files.
-			mainConf, parserConf, err := uc.GenerateFluentBitConfigs(logsDir, stateDir, hostInfo)
+func TestGenerateConfsWithValidInput(t *testing.T) {
+	for _, platform := range platforms {
+		t.Run(platform.OS, func(t *testing.T) {
+			dirPath := filepath.Join(validTestdataDir, platform.OS)
+			dirs, err := ioutil.ReadDir(dirPath)
 			if err != nil {
-				t.Fatalf("GenerateFluentBitConfigs got %v", err)
+				t.Fatal(err)
 			}
-			// Compare the expected and actual and error out in case of diff.
-			updateOrCompareGolden(t, testName, expectedMainConfig, mainConf, goldenMainPath)
-			updateOrCompareGolden(t, testName, expectedParserConfig, parserConf, goldenParserPath)
-
-			if platform == "windows" {
-				expectedOtelConfig := readFileContent(testName, goldenOtelPath, t, true)
-				otelConf, err := uc.GenerateOtelConfig(hostInfo)
-				if err != nil {
-					t.Fatalf("GenerateOtelConfig got %v", err)
-				}
-				// Compare the expected and actual and error out in case of diff.
-				updateOrCompareGolden(t, testName, expectedOtelConfig, otelConf, goldenOtelPath)
-			} else {
-				expectedCollectdConfig := readFileContent(testName, goldenCollectdPath, t, true)
-				collectdConf, err := uc.GenerateCollectdConfig(defaultLogsDir)
-				if err != nil {
-					t.Fatalf("GenerateCollectdConfig got %v", err)
-				}
-				// Compare the expected and actual and error out in case of diff.
-				updateOrCompareGolden(t, testName, expectedCollectdConfig, collectdConf, goldenCollectdPath)
+			for _, d := range dirs {
+				testName := d.Name()
+				t.Run(testName, func(t *testing.T) {
+					unifiedConfigFilePath := filepath.Join(dirPath, testName, "/input.yaml")
+					// Special-case the default config.  It lives directly in the
+					// confgenerator directory.  The golden files are still in the
+					// testdata directory.
+					if d.Name() == "default_config" {
+						unifiedConfigFilePath = platform.defaultConfig
+					}
+					
+					data, err := ioutil.ReadFile(unifiedConfigFilePath)
+					if err != nil {
+						t.Fatalf("ReadFile(%q) got %v", unifiedConfigFilePath, err)
+					}
+					uc, err := ParseUnifiedConfig(data)
+					if err != nil {
+						t.Fatalf("ParseUnifiedConfig got %v", err)
+					}
+					
+					// Retrieve the expected golden conf files.
+					expectedMainConfig := readFileContent(t, platform.OS, testName, goldenMainPath, true)
+					expectedParserConfig := readFileContent(t, platform.OS, testName, goldenParserPath, true)
+					// Generate the actual conf files.
+					setFilepathJoin(platform.OS)
```

Comment:
This kind of global usage prevents us from ever running these tests in parallel, doesn't it? Why not pass the OS into `filepathJoin` instead?

Reply:
The reason I went in that direction, was to ensure that `confgenerator` continues to use the real `filepath.Join` in real world situations. 

That said, your point is well taken. This was not a good implementation. I've reworked it to allow parallelism, but also to only override `filepath.Join` in tests.